cmd/frontend/internal/httpapi/github_webhooks_test.go

```diff
@@ -0,0 +1,253 @@
+package httpapi
+
+import (
+	"bytes"
+	"context"
+	"crypto/hmac"
+	"crypto/sha256"
+	"database/sql"
+	"encoding/hex"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"strings"
+	"testing"
+
+	gh "github.com/google/go-github/v28/github"
+
+	"github.com/sourcegraph/sourcegraph/cmd/repo-updater/repos"
+	"github.com/sourcegraph/sourcegraph/internal/db/dbtest"
+	"github.com/sourcegraph/sourcegraph/internal/db/dbtesting"
+	"github.com/sourcegraph/sourcegraph/internal/extsvc"
+	"github.com/sourcegraph/sourcegraph/schema"
+)
+
+var dsn = flag.String("dsn", "", "Database connection string to use in integration tests")
+
+func TestDispatchSuccess(t *testing.T) {
+	h := GithubWebhook{}
+	var called bool
+	h.Register("test-event-1", func(ctx context.Context, svc *repos.ExternalService, payload interface{}) error {
+		called = true
+		return nil
+	})
+
+	ctx := context.Background()
+	if err := h.Dispatch(ctx, "test-event-1", nil, nil); err != nil {
+		t.Errorf("Expected no error, got %s", err)
+	}
+	if !called {
+		t.Errorf("Expected called to be true, was false")
+	}
+}
+
+func TestDispatchNoHandler(t *testing.T) {
+	h := GithubWebhook{}
+	ctx := context.Background()
+	// no op
+	if err := h.Dispatch(ctx, "test-event-1", nil, nil); err != nil {
+		t.Errorf("Expected no error, got %s", err)
+	}
+}
+
+func TestDispatchSuccessMultiple(t *testing.T) {
+	h := GithubWebhook{}
+	var called int
+	h.Register("test-event-1", func(ctx context.Context, svc *repos.ExternalService, payload interface{}) error {
+		called++
+		return nil
+	})
+	h.Register("test-event-1", func(ctx context.Context, svc *repos.ExternalService, payload interface{}) error {
+		called++
+		return nil
+	})
+
+	ctx := context.Background()
+	if err := h.Dispatch(ctx, "test-event-1", nil, nil); err != nil {
+		t.Errorf("Expected no error, got %s", err)
+	}
+	if called != 2 {
+		t.Errorf("Expected called to be 2, got %v", called)
+	}
+}
+
+func TestDispatchError(t *testing.T) {
+	h := GithubWebhook{}
+	var called int
+	h.Register("test-event-1", func(ctx context.Context, svc *repos.ExternalService, payload interface{}) error {
+		called++
+		return fmt.Errorf("oh dear")
+	})
+	h.Register("test-event-1", func(ctx context.Context, svc *repos.ExternalService, payload interface{}) error {
+		called++
+		return nil
+	})
+
+	ctx := context.Background()
+	if err := h.Dispatch(ctx, "test-event-1", nil, nil); errString(err) != "oh dear" {
+		t.Errorf("Expected 'oh no', got %s", err)
+	}
```

Comment:
Nitpicky-nitpick because I saw "oh no" vs. "oh dear".

Here's what we often do in other parts of the codebase:

```suggestion
	if have, want := h.Dispatch(ctx, "test-event-1", nil, nil), "oh dear"; errString(have) != want {
		t.Errorf("Expected %s, got %s", want, have)
	}
```